### PR TITLE
better detection of silent samples in audio-mixer

### PR DESCRIPTION
**Pull request by zzmp**

### CODE REVIEW

**Comment by zzmp** [comment]
@kencooke, the audio-mixer is rendering a silent `float[]` into an `int16_t[]` with `1`, `-1`, and `0`, which is causing it to not send silent packets, not flush encoders, and create artifacts, especially when audio is resuming from silence.

I do not know if this is expected behavior of the audio limiter. If it is, then this PR stands. Otherwise, could you look at the audio limiter or point me in the right direction?

**Comment by hifi-gustavo** [comment]
Build console output for commit: c5424fe16dfc8cdd2bdee9931f6c373082db78bf available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-c5424fe.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-c5424fe.dmg

**Comment by ZappoMan** [comment]
Test Plan:
- Go on this server by yourself
- Enable server echo
- Enable audio noise reduction
- Turn on the audio scope (to visualize the artifacts)
- Whisper into the mic

In the current beta/master you will experience:
- When whispers break silence, you will hear artifacts.
- You should be able to see the artifacts on the received stream on the audio scope.
- They are not present on the input stream (they originate on the mixer).

On this PR you will not see/hear the artifacts.

**Comment by kencooke** [comment]
Ya that is expected. All my audio tools (limiter, resampler, reverb, etc.) have always applied TPDF dither when quantizing to int16 for the best quality. I do this mostly for unit testing, so spectrum analysis doesn't show spurious tones. But it has the effect that 0.0f input results in output of [-1,+1]

**Comment by hifi-gustavo** [comment]
Build console output for commit: 01a809e2a8ee9c9e133ce044902bf1b2263db7c3 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-01a809e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-01a809e.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 053c68b9ad9c9af6744dd48ad6dc593b46698d63 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-053c68b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-053c68b.dmg

File: assignment-client/src/audio/AudioMixerSlave.cpp [hunk 1]
```diff
@@ -221,17 +221,19 @@ bool AudioMixerSlave::prepareMix(const SharedNodePointer& listener) {
...
     for (int i = 0; i < AudioConstants::NETWORK_FRAME_SAMPLES_STEREO; ++i) {
-        if (_bufferSamples[i] != 0) {
+        if (_mixSamples[i] != 0.0f) {
             hasAudio = true;
             break;
         }
     }
+
+    // use the per listener AudioLimiter to render the mixed data
+    listenerData->audioLimiter.render(_mixSamples, _bufferSamples, AudioConstants::NETWORK_FRAME_SAMPLES_PER_CHANNEL);
```
Comment (ZappoMan) [inline assignment-client/src/audio/AudioMixerSlave.cpp]: not a blocker... but if there is `hasAudio` is false... why call ` listenerData->audioLimiter.render()`??

**Comment by cozza13** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 053c68b9ad9c9af6744dd48ad6dc593b46698d63 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-053c68b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-053c68b.dmg

**Comment by OmegaHeron** [comment]
Tested Windows 10

Master - obvious transients seen (and heard) as noise reduction threshold is cycled.

This PR (Interface and stack) - I can't discern *any* difference visually or audibly.

I also piped audio to an actual oscilloscope and spectrum analyzer to compare and I still see the heavy transients as noise reduction is broken when (very) softly speaking into microphone.

Now... let's be clear here.  I live in the subtropics of Central Florida where it's currently 80 something plus degrees and I have **lots** of white noise in background, i.e. air conditioner howling + PC fans all in close(ish) proximity to my microphone position.  Noise reduction has never ever ever worked for me in that scenario - it clicks, it pops it even goes into oscillation.  So - I'm not failing this, but, instead, passing it to others to check as well.

**Comment by ZappoMan** [comment]
I'm declaring this a QA fail... I am following up with the team. I think I have concrete evidence that this problem has existed since at least Sept... so that makes this less of a critical fix.

**Comment by zzmp** [comment]
This still fixes unexpected behavior with regards to how silence is treated by the audio-mixer, and renders certain codepaths dead that should not be. I agree that it does not fix the problem it set out to, and should no longer be considered critical.

I will add a measurement proving that it does improve the audio-mixer, and resubmit this for QA with a new test plan.

**Comment by ZappoMan** [comment]
we're close this, we found the correct fix

**Comment by zzmp** [comment]
This is an unrelated issue from #9842 - reopening - it still needs modification before it can be rereviewed.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 053c68b9ad9c9af6744dd48ad6dc593b46698d63 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-053c68b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-053c68b.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 053c68b9ad9c9af6744dd48ad6dc593b46698d63 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-053c68b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-053c68b.dmg

**Comment by zzmp** [comment]
**Test plan**

To test it you will need a client in a controlled server environment. You must use the server from this build. Any compatible client will work.
- connect client to server
- speak - observe the statistic for `avg_listeners_(silent)_per_frame` in the audio-mixer

On 972f9e3 (master, with this statistic added), it will probably be 0.
On 80b8a3f (this PR), it should go up, showing that silent packets are being properly sent.

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7e2f1a6455deb0263aced407154c3f88f7451316 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-7e2f1a6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-7e2f1a6.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 80b8a3f09cb091232488c8f26ba7b918b5deb4b0 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-80b8a3f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-80b8a3f.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 97249e31406933b83867d9fc0cc49cbb524721c7 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-97249e3.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-97249e3.dmg

**Comment by ZappoMan** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7e2f1a6455deb0263aced407154c3f88f7451316 available at: https://gist.github.com/4d93fd42e735bf52a02fd303a928a69a<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-7e2f1a6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9830-7e2f1a6.dmg

**Comment by OmegaHeron** [comment]
Windows 10

Using this PR stack
Watched stat specified.
It remains 1 with 1 user connected and mic muted.
Mic not muted, noise reduction on, not breaking noise threshold, it remains 1.
Mic not muted, noise reduction off, it remains 0.

I'll buy it - I'm convinced with noise reduction on, AC is getting it has silent frames.

**Pull request merged**